File: backend/app/services/forecast.py
"""Projeccio del saldo i deteccio anticipada de descoberts.

La projeccio suma al saldo actual els rebuts recurrents previstos i una deriva
de despesa no recurrent estimada a partir dels ultims mesos. Es dona en tres
linies (esperada, optimista i pessimista) perque la despesa variable no es
previsible amb una sola xifra.
"""

from __future__ import annotations

import logging
from dataclasses import dataclass, field
from datetime import date, timedelta
from decimal import Decimal

from sqlalchemy import select
from sqlalchemy.orm import Session

from app.config import settings
from app.core.time import today_local
from app.models import Ledger, RecurringOccurrence, RecurringSeries, Transaction
from app.models.enums import (
    AlertSeverity,
    AlertType,
    SeriesStatus,
    TransactionStatus,
)
from app.services.alerts import create_alert
from app.services.balances import ledger_balance

logger = logging.getLogger(__name__)

# Historic que es mira per estimar la despesa variable.
DISCRETIONARY_WINDOW_DAYS = 90
# Els imports mes grans es descarten: una compra excepcional no es una tendencia.
OUTLIER_TRIM_RATIO = 0.05
# Amplada de la banda optimista i pessimista sobre la despesa variable.
BAND_SPREAD = Decimal("0.30")


@dataclass
class ForecastEvent:
    day: date
    label: str
    amount: Decimal
    series_id: int | None = None


@dataclass
class ForecastPoint:
    day: date
    expected: Decimal
    optimistic: Decimal
    pessimistic: Decimal


@dataclass
class Forecast:
    ledger_id: int
    ledger_name: str
    currency: str
    starting_balance: Decimal
    threshold: Decimal
    horizon_days: int
    daily_discretionary: Decimal
    points: list[ForecastPoint] = field(default_factory=list)
    events: list[ForecastEvent] = field(default_factory=list)
    first_breach_day: date | None = None
    first_breach_amount: Decimal | None = None


def daily_discretionary_spend(db: Session, ledger_id: int) -> Decimal:
    """Despesa diaria mitjana que no ve de cap rebut recurrent.

    Es descarten els imports mes alts perque una compra excepcional no marqui
    la tendencia de tot el trimestre.
    """
    since = today_local() - timedelta(days=DISCRETIONARY_WINDOW_DAYS)
    recurring_ids = set(
        db.scalars(
            select(RecurringOccurrence.transaction_id)
            .join(RecurringSeries, RecurringSeries.id == RecurringOccurrence.series_id)
            .where(RecurringSeries.ledger_id == ledger_id)
        )
    )

    amounts = [
        -transaction.amount
        for transaction in db.scalars(
            select(Transaction).where(
                Transaction.ledger_id == ledger_id,
                Transaction.booking_date >= since,
                Transaction.amount < 0,
                Transaction.status == TransactionStatus.BOOKED,
                Transaction.transfer_group_id.is_(None),
                Transaction.is_excluded.is_(False),
            )
        )
        if transaction.id not in recurring_ids
    ]
    if not amounts:
        return Decimal("0.00")

    amounts.sort()
    trim = int(len(amounts) * OUTLIER_TRIM_RATIO)
    if trim:
        amounts = amounts[:-trim]
    total = sum(amounts, Decimal("0.00"))
    return (total / Decimal(DISCRETIONARY_WINDOW_DAYS)).quantize(Decimal("0.01"))


def upcoming_events(
    db: Session, ledger_id: int, horizon: date, start: date | None = None
) -> list[ForecastEvent]:
    """Rebuts recurrents previstos d'aqui a l'horitzo."""
    start = start or today_local()
    events: list[ForecastEvent] = []

    for series in db.scalars(
        select(RecurringSeries).where(
            RecurringSeries.ledger_id == ledger_id,
            RecurringSeries.status == SeriesStatus.ACTIVE,
            RecurringSeries.include_in_forecast.is_(True),
        )
    ):
        interval = max(series.interval_days, 1)
        occurrence = series.next_expected_date or (series.last_seen_date + timedelta(interval))
        # Si la data prevista ja ha passat, s'avanca fins a la primera futura.
        while occurrence < start:
            occurrence += timedelta(days=interval)
        while occurrence <= horizon:
            events.append(
                ForecastEvent(
                    day=occurrence,
                    label=series.label,
                    amount=series.expected_amount,
                    series_id=series.id,
                )
            )
            occurrence += timedelta(days=interval)

    events.sort(key=lambda item: item.day)
    return events


def build_forecast(db: Session, ledger: Ledger, horizon_days: int | None = None) -> Forecast:
    horizon_days = horizon_days or settings.forecast_horizon_days
    start = today_local()
    horizon = start + timedelta(days=horizon_days)

    balance, _ = ledger_balance(db, ledger.id)
    daily = daily_discretionary_spend(db, ledger.id)
    events = upcoming_events(db, ledger.id, horizon, start)

    forecast = Forecast(
        ledger_id=ledger.id,
        ledger_name=ledger.name,
        currency=ledger.currency,
        starting_balance=balance,
        threshold=ledger.overdraft_threshold,
        horizon_days=horizon_days,
        daily_discretionary=daily,
        events=events,
    )

    events_by_day: dict[date, Decimal] = {}
    for event in events:
        events_by_day[event.day] = events_by_day.get(event.day, Decimal("0.00")) + event.amount

    running = balance
    optimistic_drift = daily * (Decimal("1") - BAND_SPREAD)
    pessimistic_drift = daily * (Decimal("1") + BAND_SPREAD)

    for offset in range(horizon_days + 1):
        day = start + timedelta(days=offset)
        running += events_by_day.get(day, Decimal("0.00"))
        expected = (running - daily * offset).quantize(Decimal("0.01"))
        point = ForecastPoint(
            day=day,
            expected=expected,
            optimistic=(running - optimistic_drift * offset).quantize(Decimal("0.01")),
            pessimistic=(running - pessimistic_drift * offset).quantize(Decimal("0.01")),
        )
        forecast.points.append(point)

        if forecast.first_breach_day is None and expected < ledger.overdraft_threshold:
            forecast.first_breach_day = day
            forecast.first_breach_amount = expected

    return forecast


def check_overdrafts(db: Session, horizon_days: int | None = None) -> int:
    """Genera un avis per cada llibre que es preveu que entri en descobert."""
    created = 0
    for ledger in db.scalars(select(Ledger).where(Ledger.is_active.is_(True))):
        forecast = build_forecast(db, ledger, horizon_days)
        if forecast.first_breach_day is None:
            continue

        days_ahead = (forecast.first_breach_day - today_local()).days
        cause = next(
            (
                event
                for event in forecast.events
                if event.day <= forecast.first_breach_day and event.amount < 0
            ),
            None,
        )
        body = (
            f"Amb el saldo actual de {forecast.starting_balance:.2f} EUR, els rebuts "
            f"previstos i una despesa variable de {forecast.daily_discretionary:.2f} EUR al "
            f"dia, el saldo baixaria a {forecast.first_breach_amount:.2f} EUR el "
            f"{forecast.first_breach_day:%d/%m/%Y}."
        )
        if cause is not None:
            body += f" El primer rebut important previst es {cause.label}."

        alert = create_alert(
            db,
            type=AlertType.PROJECTED_OVERDRAFT,
            ledger_id=ledger.id,
            # Una alerta per llibre i setmana: el mateix descobert no ha d'avisar cada dia.
            dedup_key=f"overdraft:{ledger.id}:{forecast.first_breach_day.isocalendar()[:2]}",
            title=f"{ledger.name}: possible descobert d'aqui a {days_ahead} dies",
            body=body,
            severity=AlertSeverity.CRITICAL if days_ahead <= 14 else AlertSeverity.WARNING,
            payload={
                "ledger_id": ledger.id,
                "breach_day": forecast.first_breach_day.isoformat(),
                "breach_amount": str(forecast.first_breach_amount),
                "starting_balance": str(forecast.starting_balance),
            },
        )
        if alert is not None:
            created += 1
    db.flush()
    return created

File: src/services/forecast.ts
/**
 * Projection of the balance from the confirmed expected direct debits.
 *
 * The active schedules with `include_in_forecast` are added to today's
 * balance. There is no residual «variable expense»: what is not an expected
 * direct debit is looked at in the reports, not in the forecast.
 */

import { and, eq } from "drizzle-orm";

import { db } from "../db/client.ts";
import { recurringSeries, type Ledger } from "../db/schema/index.ts";
import { config } from "../lib/config.ts";
import { Decimal, money, toMoneyString, ZERO, type MoneyString } from "../lib/money.ts";
import { addDays, daysBetween, todayLocal } from "../lib/time.ts";
import { createAlert } from "./alerts.ts";
import { workspaceBalance, balanceSeries, type BalancePoint } from "./balances.ts";

export interface EventExpected {
  day: string;
  label: string;
  amount: MoneyString;
  seriesId: number | null;
}

export interface ForecastPoint {
  day: string;
  expected: MoneyString;
  optimista: MoneyString;
  pessimistic: MoneyString;
  /** Least-squares line over `esperat`: the overall trend. */
  trend: MoneyString;
}

export interface Forecast {
  ledgerId: number;
  ledgerName: string;
  currency: string;
  openingBalance: MoneyString;
  threshold: MoneyString;
  horizonDays: number;
  /** Always zero: kept in the type so as not to break the charts' UI. */
  dailySpend: MoneyString;
  /** Real balance rebuilt backwards (same width as the horizon). */
  history: BalancePoint[];
  points: ForecastPoint[];
  events: EventExpected[];
  firstOverdraft: string | null;
  firstOverdraftAmount: MoneyString | null;
}

/** Confirmed expected direct debits from here to the horizon. */
export async function eventsExpected(
  ledgerId: number,
  horitzo: string,
  start?: string,
): Promise<EventExpected[]> {
  const begin = start ?? todayLocal();
  const events: EventExpected[] = [];

  const activeSeries = await db
    .select()
    .from(recurringSeries)
    .where(
      and(
        eq(recurringSeries.ledgerId, ledgerId),
        eq(recurringSeries.status, "active"),
        eq(recurringSeries.includeInForecast, true),
      ),
    );

  for (const series of activeSeries) {
    const interval = Math.max(series.intervalDays, 1);
    let occurrence = series.nextExpectedDate ?? addDays(series.lastSeenDate, interval);

    while (occurrence < begin) occurrence = addDays(occurrence, interval);

    while (occurrence <= horitzo) {
      events.push({
        day: occurrence,
        label: series.label,
        amount: series.expectedAmount,
        seriesId: series.id,
      });
      occurrence = addDays(occurrence, interval);
    }
  }

  events.sort((a, b) => a.day.localeCompare(b.day));
  return events;
}

export async function buildForecast(
  workspace: Ledger,
  horizonDays?: number,
): Promise<Forecast> {
  const days = horizonDays ?? config.forecastHorizonDays;
  const start = todayLocal();
  const horitzo = addDays(start, days);

  const [{ total: balance }, events, history] = await Promise.all([
    workspaceBalance(workspace.id),
    eventsExpected(workspace.id, horitzo, start),
    // Same width to the left and to the right of the chart.
    balanceSeries([workspace.id], addDays(start, -days), start),
  ]);

  const byDay = new Map<string, Decimal>();
  for (const e of events) {
    byDay.set(e.day, (byDay.get(e.day) ?? new Decimal(0)).plus(money(e.amount)));
  }

  const threshold = money(workspace.overdraftThreshold);

  const pointsWithoutTrend: Omit<ForecastPoint, "trend">[] = [];
  let running = money(balance);
  let firstOverdraft: string | null = null;
  let firstOverdraftAmount: MoneyString | null = null;

  for (let offset = 0; offset <= days; offset += 1) {
    const day = addDays(start, offset);
    running = running.plus(byDay.get(day) ?? new Decimal(0));
    const expected = running.toDecimalPlaces(2);

    // With no residual expense, the bands coincide with the expected value.
    pointsWithoutTrend.push({
      day,
      expected: toMoneyString(expected),
      optimista: toMoneyString(expected),
      pessimistic: toMoneyString(expected),
    });

    if (firstOverdraft === null && expected.lt(threshold)) {
      firstOverdraft = day;
      firstOverdraftAmount = toMoneyString(expected);
    }
  }

  const trendLine = leastSquaresLine(pointsWithoutTrend.map((p) => money(p.expected)));
  const points: ForecastPoint[] = pointsWithoutTrend.map((p, i) => ({
    ...p,
    trend: toMoneyString(trendLine[i] ?? ZERO),
  }));

  return {
    ledgerId: workspace.id,
    ledgerName: workspace.name,
    currency: workspace.currency,
    openingBalance: balance,
    threshold: workspace.overdraftThreshold,
    horizonDays: days,
    dailySpend: "0.00",
    history,
    points,
    events,
    firstOverdraft,
    firstOverdraftAmount,
  };
}

export function leastSquaresLine(values: Decimal[]): Decimal[] {
  const n = values.length;
  if (n === 0) return [];
  if (n === 1) return [values[0] ?? ZERO];

  let sumX = ZERO;
  let sumY = ZERO;
  let sumXY = ZERO;
  let sumXX = ZERO;

  for (let i = 0; i < n; i += 1) {
    const x = new Decimal(i);
    const y = values[i] ?? ZERO;
    sumX = sumX.plus(x);
    sumY = sumY.plus(y);
    sumXY = sumXY.plus(x.times(y));
    sumXX = sumXX.plus(x.times(x));
  }

  const nDec = new Decimal(n);
  const denominator = nDec.times(sumXX).minus(sumX.times(sumX));
  if (denominator.isZero()) {
    const average = sumY.dividedBy(nDec);
    return values.map(() => average.toDecimalPlaces(2));
  }

  const slope = nDec.times(sumXY).minus(sumX.times(sumY)).dividedBy(denominator);
  const origin = sumY.minus(slope.times(sumX)).dividedBy(nDec);

  return values.map((_, i) => origin.plus(slope.times(i)).toDecimalPlaces(2));
}

function isoWeek(isoDate: string): string {
  const [y, m, d] = isoDate.split("-").map(Number);
  const date = new Date(Date.UTC(y ?? 1970, (m ?? 1) - 1, d ?? 1));
  const day = date.getUTCDay() || 7;
  date.setUTCDate(date.getUTCDate() + 4 - day);
  const start = new Date(Date.UTC(date.getUTCFullYear(), 0, 1));
  const week = Math.ceil(((date.getTime() - start.getTime()) / 86_400_000 + 1) / 7);
  return `${date.getUTCFullYear()}-${week}`;
}

export async function checkOverdrafts(
  workspace: Ledger,
  horizonDays?: number,
): Promise<number> {
  const forecast = await buildForecast(workspace, horizonDays);
  if (forecast.firstOverdraft === null) return 0;

  const viewDays = daysBetween(todayLocal(), forecast.firstOverdraft);
  const cause = forecast.events.find(
    (e) => e.day <= (forecast.firstOverdraft as string) && money(e.amount).isNegative(),
  );

  let body =
    `Amb el saldo actual de ${money(forecast.openingBalance).toFixed(2)} EUR i els rebuts ` +
    `previstos confirmats, el saldo baixaria a ${money(forecast.firstOverdraftAmount).toFixed(2)} EUR el ` +
    `${forecast.firstOverdraft}.`;
  if (cause) body += ` El primer rebut important previst es ${cause.label}.`;

  const createdOne = await createAlert({
    type: "projected_overdraft",
    ledgerId: workspace.id,
    dedupKey: `overdraft:${workspace.id}:${isoWeek(forecast.firstOverdraft)}`,
    title: `${workspace.name}: possible descobert d'aqui a ${viewDays} dies`,
    body: body,
    severity: viewDays <= 14 ? "critical" : "warning",
    payload: {
      ledger_id: workspace.id,
      breach_day: forecast.firstOverdraft,
      breach_amount: forecast.firstOverdraftAmount,
      starting_balance: forecast.openingBalance,
    },
  });

  return createdOne ? 1 : 0;
}
